### ISSUE TITLE
[DOC] Question optimization criterion

### ISSUE THREAD

**Issue by mmikhasenko** [arogozhnikov/hep_ml#89]
Hi @arogozhnikov,
Thanks for the cool package. I came across the GBReweighter a few times, it's actively used in LHCb to these days.

Recently, I looked at the code trying to find more about internals. I was looking for the **symmetrized binned chi-squared** 
<img width="729" alt="Image" src="https://github.com/user-attachments/assets/156e2dc1-966c-4502-a991-7232a9cd13b5" />
that mentioned in several places ([paper](https://arxiv.org/pdf/1608.05806), [blog](https://arogozhnikov.github.io/2015/10/09/gradient-boosted-reweighter.html), [lecture](https://indico.cern.ch/event/439520/contributions/1941491/attachments/1146711/1644246/lecture4.pdf)) to be the optimization objective

but I could not figure out how it comes about in this form.
Could you please help?

## What's in the code

I looks to me that objective is computed as
```
leaf_prediction = weighted_average(signs in leaf)
MSE = sum(weights * (signs - leaf_prediction)^2)
```

[Here](https://github.com/arogozhnikov/hep_ml/blob/fa08301c7fe975c4f8f53bf997abf63ae40a40c9/hep_ml/gradientboosting.py#L134-L135) is the code falls back to the `SklearnClusteringTree` with MSE

```python
tree = SklearnClusteringTree(
                criterion='squared_error',
...
...
tree.fit(X[train_indices], residual[train_indices],
          sample_weight=weights[train_indices], check_input=False)
```
where the `residual` refers to the `signs`.

Such loss function would try to separate `original` and `target` identifying regions of the largest mismatch.

## Math

The leaf value that minimizes the MSE is a weighted sum,

$$v = \frac{\sum w_i^o - \sum w_i^t}{n}$$

where $n = \sum w_i^o + \sum w_i^t$ is the sum

Then, I can simplify the weighed squared error, 
```math
\text{loss} = \sum_i w_i (s_i - v)^2 = n (1-v^2)
```

that is maximum when $v=\pm 1$ - all points in the leaf are of the same class.

**Comment by arogozhnikov** [comment]
> it's actively used in LHCb to these days.

glad to hear that.

There was a question (https://github.com/arogozhnikov/hep_ml/issues/32), but it seems you want more detailed derivation

Let's consider a split of distribution into two bins, left and right, introduce $W^t_l, W^t_r$ for total weight of target events in each bin, and $W^o_l, W^o_r$ for total weight of original events in each bin (just to avoid writing $\Sigma$ ).

Tree optimizes MSE loss, you correctly read that targets are signs (+1, -1), and weights correspond to weight of particular sample (event); total MSE for two bins takes a form:

```math
MSE = W^t_l (v_l - 1)^2 + W^o_l (v_l + 1)^2  + W^t_r (v_r - 1)^2 + W^o_r (v_r + 1)^2
```

which, after minimizing over values in leaves `v_l`, `v_r` turns into (important exercise: do the math here by computing derivative, optimum values in general case are *not* +1 or -1) 

```math
MSE = \frac{4 W^t_l  W^o_l}{W^t_l + W^o_l}  +  \frac{4 W^t_r  W^o_r}{W^t_r + W^o_r}
```

then we subtract total weight of all events (this value does not depend on the split, and effectively is an external constant for the optimization)

```math
MSE - (W^t_l + W^o_l + W^t_r + W^o_r)  = \frac{4 W^t_l  W^o_l - (W^t_l + W^o_l)^2}{W^t_l + W^o_l}  +  \frac{4 W^t_r  W^o_r - (W^t_r + W^o_r)^2}{W^t_r + W^o_r}
```
simple transition:
```math
MSE - (W^t_l + W^o_l + W^t_r + W^o_r)  = \frac{- (W^t_l - W^o_l)^2}{W^t_l + W^o_l}  +  \frac{- (W^t_r - W^o_r)^2}{W^t_r + W^o_r}
```
so what we get is 
```math
MSE - \text{ (some const not depending on selected split) }  = - \chi^2
```
and so split that minimizes MSE, also maximizes $\chi^2$. Tree optimization from sklearn is used only to find these splits, computed leaf values are discarded.

**Comment by arogozhnikov** [comment]
>  is there anything that makes the first step different from a binary classification using MSE?

tree structure should be identical (selected features and thresholds), values in leaves are of course different.

**Comment by mmikhasenko** [comment]
Many thanks, resolved
